lib/Dialect/LWE/IR/NewLWEAttributes.td

```diff
@@ -0,0 +1,352 @@
+}
+
+def LWE_KeyAttr : AttrDef<LWE_Dialect, "Key"> {
+  let mnemonic = "key";
+  let description = [{
+    An attribute describing the key used for encrypting the ciphertext.
+
+    This attribute includes a key identifier for the original key used to
+    encrypt the secret key.
+
+    The `key_size` parameter is used to describe the number of polynomials of
+    the secret key. This is typically $1$ for RLWE ciphertexts and greater than
+    $1$ for LWE instances. A ciphertext encrypted with a `key_size` of $k$ will
+    have size $k+1$. This is the size of the mask.
+
+    The key basis describes the inner product used in the phase calculation in
+    decryption. This attribute is only supported for RLWE ciphertexts whose
+    `key_size` is $1$. An RLWE ciphertext is canonically encrypted against key
+    basis `(1, s)`. After a multiplication, its size will increase and the basis
+    will be `(1, s, s^2)`. The array that represents the key basis is
+    constructed by listing the powers of `s` at each position of the array. For
+    example, `(1, s, s^2)` corresponds to `[0, 1, 2]`, while `(1, s^2)`
+    corresponds to `[0, 2]`.
+  }];
+
+  let parameters = (ins
+    "::mlir::StringAttr":$id,
+    DefaultValuedParameter<"unsigned", "1">:$size,
+    OptionalArrayRefParameter<"unsigned int">:$basis
+  );
+
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+def LWE_EncryptionTypeEnum : I32EnumAttr<"LweEncryptionType", "An enum attribute representing an encryption method", [
+  I32EnumAttrCase<"msb", 0>,
+  I32EnumAttrCase<"lsb", 1>,
+  I32EnumAttrCase<"mix", 2>
+]> {
+    let cppNamespace = "::mlir::heir::lwe";
+}
+
+def LWE_CiphertextSpaceAttr : AttrDef<LWE_Dialect, "CiphertextSpace"> {
+  let mnemonic = "ciphertext_space";
+  let description = [{
+    An attribute describing the ciphertext space and the transformation from
+    plaintext space to ciphertext space of an FHE scheme.
+
+    The ciphertext space information includes the ring structure, which contains
+    the ciphertext modulus $q$. Ciphertexts using an RNS representation for $q$
+    will represent their ciphertext components in the ring attribute. Scalar LWE
+    ciphertexts (as opposed to RLWE) use an ideal polynomial of degree 1, $x$.
+    CGGI ciphertexts will typically use a power of two modulus.
+
+    The ciphertext encoding info is used to describe the way the plaintext data
+    is encoded into the ciphertext (in the MSB, LSB, or mixed).
+  }];
+
+  let parameters = (ins
+    "::mlir::polynomial::RingAttr":$ciphertext_ring,
+    "::mlir::heir::lwe::LweEncryptionType":$encryption_type
+  );
+
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+def LWE_ModulusChainAttr : AttrDef<LWE_Dialect, "ModulusChain"> {
+  let mnemonic = "modulus_chain";
+  let description = [{
+    An attribute describing the modulus chain of an RLWE ciphertext.
+  }];
+
+  let parameters = (ins
+    ArrayRefParameter<"mlir::IntegerAttr">:$chain,
```

Comment:
```suggestion
    ArrayRefParameter<"mlir::IntegerAttr">:$moduli,
```

Reply:
Or perhaps, "elements" as these are the elements of the chain. It is needlessly self-referential to say that a "chain" is defined by containing a "chain".